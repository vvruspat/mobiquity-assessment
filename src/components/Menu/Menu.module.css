.menu {
    width: 100%;
    height: 36px;

    overflow-x: scroll;
    scroll-snap-type: x mandatory;
    scrollbar-width: none;
    
    display: flex;
    flex-direction: row;
    align-items: center;
    gap: var(--padding-regular);

    box-sizing: border-box;
    padding: 0 var(--padding-large);
}

.menu::-webkit-scrollbar {
    display: none;
}

.menu_item {
    scroll-snap-align: center;
}