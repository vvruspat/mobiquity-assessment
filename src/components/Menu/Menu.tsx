import { FC, PropsWithChildren, useCallback, useEffect, useState } from "react";
import { Button } from "../UIKit/Button";

import s from "./Menu.module.css";

export type MenuItemOption = {
  /** Key to identify option */
  key: string;
  /** Value to show */
  value: string;
};

type MenuProps = {
  /** Menu options */
  options: MenuItemOption[];
  /** Allow to select item from outside */
  selectedIndex?: number;
  /** Callback when item has been selected */
  onSelect?: (selectedItem: MenuItemOption) => void;
};

/**
 * Menu component
 */
export const Menu: FC<PropsWithChildren<MenuProps>> = ({
  options,
  selectedIndex = 0,
  onSelect,
}) => {
  const [selectedKey, setSelectedKey] = useState(options[selectedIndex]?.key);
  const onMenuItemClick = useCallback(
    (selectedItem: MenuItemOption) => {
      onSelect?.(selectedItem);
      setSelectedKey(selectedItem.key);
    },
    [onSelect]
  );

  useEffect(() => {
    if (options[selectedIndex]) {
      setSelectedKey(options[selectedIndex].key);
    }
  }, [selectedIndex, options]);

  return (
    <div className={s.menu} dir="ltr" data-testid="menu">
      {options.map((option) => {
        return (
          <Button
            mode={selectedKey === option.key ? "primary" : "outline"}
            key={option.key}
            onClick={() => onMenuItemClick(option)}
            className={s.menu_item}
          >
            {option.value}
          </Button>
        );
      })}
    </div>
  );
};
