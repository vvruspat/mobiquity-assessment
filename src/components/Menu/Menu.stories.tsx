import { ComponentStory, ComponentMeta } from "@storybook/react";

import { Menu } from "./Menu";
import "../../index.css";

export default {
  title: "Components/Menu",
  component: Menu,
} as ComponentMeta<typeof Menu>;

const Template: ComponentStory<typeof Menu> = (args) => (
  <Menu {...args}>Menu content</Menu>
);

export const MenuTemplate = Template.bind({});

const options = [];

for (let i = 0; i < 20; i++) {
  const year = (2022 - Number(i)).toString();

  options.push({
    key: year,
    value: year,
  });
}

MenuTemplate.args = {
  options,
};
