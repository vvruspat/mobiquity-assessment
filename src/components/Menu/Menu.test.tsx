import { render, screen } from "@testing-library/react";
import { Menu } from "./Menu";

test("renders menu", () => {
  render(
    <Menu
      options={[
        { key: "1", value: "1 text" },
        { key: "2", value: "2 text" },
      ]}
    />
  );

  const el = screen.getByTestId("menu");
  expect(el).toMatchSnapshot();

  const buttons = screen.getAllByTestId("button");
  expect(buttons.length).toEqual(2);
  expect(buttons[0]).toHaveTextContent("1 text");
  expect(buttons[1]).toHaveTextContent("2 text");
});
