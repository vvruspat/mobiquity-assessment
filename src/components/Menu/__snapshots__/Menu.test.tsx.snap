// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders menu 1`] = `
<div
  class="menu"
  data-testid="menu"
  dir="ltr"
>
  <button
    class="button primary-button menu_item"
    data-testid="button"
  >
    1 text
  </button>
  <button
    class="button outline-button menu_item"
    data-testid="button"
  >
    2 text
  </button>
</div>
`;
